integrations/clickup/client.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import time
+from typing import Any, AsyncGenerator, Optional, List, Dict
+from httpx import HTTPStatusError, Timeout
+from loguru import logger
+from port_ocean.utils import http_async_client
+from port_ocean.utils.cache import cache_iterator_result
+
+TEAM_OBJ = "__team"
+WEBHOOK_EVENTS = [
+    "taskCreated",
+    "taskUpdated",
+    "taskDeleted",
+    "listCreated",
+    "listUpdated",
+    "listDeleted",
+]
+MINIMUM_LIMIT_REMAINING = 20
+DEFAULT_SLEEP_TIME = 30
+
+# Adjust the concurrency level as needed
+SEMAPHORE = asyncio.BoundedSemaphore(10)
+
+
+class ClickupClient:
+    """Clickup client to interact with Clickup API."""
+
+    def __init__(self, clickup_url: str, clickup_token: str, archived: bool):
+        self.clickup_url = clickup_url
+        self.clickup_token = clickup_token
+        self.api_url = f"{self.clickup_url}/api/v2"
+        self.client = http_async_client
+        self.client.timeout = Timeout(60)
+        self.archived = archived
+
+    @property
+    def api_headers(self) -> dict[str, Any]:
+        return {
+            "Authorization": self.clickup_token,
+            "Content-Type": "application/json",
+        }
+
+    async def _fetch_with_rate_limit_handling(
+        self,
+        url: str,
+        method: str,
+        params: Optional[dict[str, Any]] = None,
+        json_data: Optional[dict[str, Any]] = None,
+    ) -> Any:
+        """
+        Sends an HTTP request to the ClickUp API with rate limit handling.
+        """
+        while True:
+            async with SEMAPHORE:
+                response = await self.client.request(
+                    url=url,
+                    method=method,
+                    headers=self.api_headers,
+                    params=params,
+                    json=json_data,
+                )
+                try:
+                    response.raise_for_status()
+                    rate_limit_remaining = int(
+                        response.headers.get("X-RateLimit-Remaining", 1)
+                    )
+                    rate_limit_reset = int(response.headers.get("X-RateLimit-Reset"))
+                    if rate_limit_remaining <= MINIMUM_LIMIT_REMAINING:
+                        current_time = int(time.time())
+                        wait_time = max(
+                            rate_limit_reset - current_time, DEFAULT_SLEEP_TIME
+                        )
```

Comment:
Include reference to where one can get information on Clickup rate limit parameters and the duration

Reply:
https://clickup.com/api/developer-portal/rate-limits/

The documentation link has been included for reference

Reply:
I followed the link but I don't see where it mentions the `X-RateLimit-Rese`t header

Reply:
@oiadebayo please update the link to where its states about the rate limit headers

Reply:
I have included additional link where the header parameter was mentioned